lib/teleterm/clusters/cluster_headless.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clusters
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+)
+
+// WatchPendingHeadlessAuthentications watches the backend for pending headless authentication requests for the user.
+func (c *Cluster) WatchPendingHeadlessAuthentications(ctx context.Context) (watcher types.Watcher, close func(), err error) {
+	err = addMetadataToRetryableError(ctx, func() error {
+		proxyClient, err := c.clusterClient.ConnectToProxy(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		rootClient, err := proxyClient.ConnectToRootCluster(ctx)
+		if err != nil {
+			proxyClient.Close()
+			return trace.Wrap(err)
+		}
+
+		watcher, err = rootClient.WatchPendingHeadlessAuthentications(ctx)
+		if err != nil {
+			proxyClient.Close()
+			rootClient.Close()
+			return trace.Wrap(err)
+		}
+
+		close = func() {
+			watcher.Close()
+			proxyClient.Close()
+			rootClient.Close()
+		}
+
+		return nil
+	})
+	return watcher, close, trace.Wrap(err)
+}
+
+// UpdateHeadlessAuthenticationState updates the headless authentication matching the given id to the given state.
+// MFA will be prompted when updating to the approve state.
+func (c *Cluster) UpdateHeadlessAuthenticationState(ctx context.Context, headlessID string, state types.HeadlessAuthenticationState) error {
+	err := addMetadataToRetryableError(ctx, func() error {
+		proxyClient, err := c.clusterClient.ConnectToProxy(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer proxyClient.Close()
+
+		rootClient, err := proxyClient.ConnectToRootCluster(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer rootClient.Close()
+
+		// If changing state to approved, create an MFA challenge and prompt for MFA.
+		var mfaResponse *proto.MFAAuthenticateResponse
+		if state == types.HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_APPROVED {
+			chall, err := rootClient.CreateAuthenticateChallenge(ctx, &proto.CreateAuthenticateChallengeRequest{
+				Request: &proto.CreateAuthenticateChallengeRequest_ContextUser{
+					ContextUser: &proto.ContextUser{},
+				},
+			})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			mfaResponse, err = c.clusterClient.PromptMFAChallenge(ctx, c.clusterClient.WebProxyAddr, chall, nil)
```

Comment:
This is related to my comment from the other PR, https://github.com/gravitational/teleport/pull/29097#discussion_r1266772066. This is going to use the default CLI-based terminal prompt which is going to break the flow if the authenticator ever asks for a PIN during MFA.

https://github.com/gravitational/teleport/blob/fa6a2982200b98c228cb447815f70c94347cabca/lib/client/mfa.go#L248-L251

https://github.com/gravitational/teleport/blob/fa6a2982200b98c228cb447815f70c94347cabca/lib/client/mfa.go#L41-L48

The comment says it shouldn't really happen. I suppose it'd be best to have a prompt implementation which returns `NotImplemented` in this case instead of making the app hang. Alas, it looks like it'd require refactoring a lot of `PromptMFAChallenge`.

So at this point I guess we just cross fingers that it never happens?

BTW, does headless auth require per-session MFA to be enabled on the cluster?

Reply:
I'm going to leave it out of scope for now. Refactoring `PromptMFAChallenge` is [in my near future](https://github.com/gravitational/teleport/pull/27588), so I'll keep this in mind.

Reply:
Also, headless does not require per-session MFA to be enabled, only WebAuthn.

Reply:
Now that I think about it, I'm pretty sure we will never prompt for PIN. We only use the Passwordless flow for login (IIUC Passwordless basically does WebAuthn + some). For MFA we use Webauthn directly.